Exception-Handling.md

```diff
@@ -0,0 +1,209 @@
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+rescue
+  # other exceptions handled here
+
+ensure
+  #  This code runs no matter what.  
+
+end 
+```
+#### One Ruby Example - CSVReader
+
+In Ruby we could write code to handle problems, like in our CSV code to deal with problems, like being unable to open the CSV file.
+
+```ruby
+content = []
+begin
+  CSV.foreach("file.csv") do |row|
+
+    content << row
+  end
+
+  rescue SystemCallError
+    puts "I couldn't open the file."
+
+end
+```
+
+#### Exceptions Continue Up The Stack
+
+When an exception occurs, if it is not rescued in the immediate block or method it proceeds up and out of the method until the application crashes with an error message.
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << quotient % divisor
+  answer
+end
+
+begin
+  puts quotient_and_remainer 10, 0
+rescue ZeroDivisionError
+  puts "Attempted to divide by zero"
+end
+```
+The method `divide` here will generate an exception when it tries to divide 10 by zero.  The exception is not "caught" in the method so ruby leaves the method and falls back to the caller of the method, which does have a rescue block to catch the `ZeroDivisionError`.  
+
+
+```bash
+$  ruby example.rb
+Attempted to divide by zero
+```
+
+
+#### Exception Handling in Rails
+
+In Rails we have a number of methods in ActiveRecord which can generate exceptions a few of which include:
+-  save!
+-  create!
+-  update!
+
+These methods trigger exceptions when validations fail or the database is unavailable.  So you could handle problems with creating a new task in TaskListRails using exceptions as follows.  
+
+```ruby
+  def create
+    @task = Task.new(task_params)
+
+    begin
+      @task.save!
+    rescue ActiveRecord::RecordInvalid => e
+      msg = ""
+      e.record.errors.each do |field, message |
+        msg += "#{field.to_s.capitalize}: #{message}"
+      end
+      flash[:notice] = msg
+    ensure
+      redirect_to tasks_new_path
+    end
+  end
+```
+
+However while this is simple it isn't good practice.  It's considered best practice to [**not use exceptions for expected outcomes**](https://robots.thoughtbot.com/save-bang-your-head-active-record-will-drive-you-mad).  Since we expect users to give us invalid input on a regular basis, the above example does not qualify.  This leaves us with the question:  
```

Comment:
I think we should avoid calling the above code simple. I'm not sure what an alternative adjective would be, other than maybe "possible".